DESIGN.md

```diff
@@ -0,0 +1,168 @@
+
+Each step is responsible for error handling. In the case that a step is responsible for creation of just one resource (recommended) there is usually no need to add additional error handling,
+but this can be changed depending on complexity/requirements. The base step from which all steps are inherited contains some common error handling as well;
+for instance, the base class handles the situation when the resource already exists.
+
+## Deleting Landing Zones
+
+Deleting resources in a landing zone presents several challenges:
+
+- Order of deletion is important. Many resources have a hard dependency on each other, and you can't delete a resource without deleting its dependent resource first. Unfortunately, the ARM API does not expose an easy way to list all these dependencies. Therefore the implementation must identify each dependency case explicitly.
+
+- Guarantee safety when deleting the landing zone resources. Resources in the workspace could use the resources in the landing zone, and when this is the case, you can't delete the landing zone. The definition of "usage" is not consistent and is specific to the resource. For example, an Azure Relay namespace is in use if a hybrid connection exists; in contrast, a VNet is considered in use if there's a VM attached to one of its subnets.
+
+- Landing zone definitions are extendable. Landing zones can be modified with new resources or definitions. Therefore, resource deletion must be flexible to address future scenarios.
+
+### Deletion Order
+
+To address the deletion order, the implementation groups the resource into two groups. The first group is the dependent resources, and the second is the base resources. Base resources are deleted last.
+
+The enum `LandingZoneBaseResourceType` contains the list of base resources, currently on a VNet is considered a base resource. As definitions evolved, additional base resources can be added to the enum.
+
+Private endpoints are a special case. They must be deleted before a dependent and the base resources. The implementation identifies all the landing zone resources with a private endpoint to handle this case.
+
+### Deletion Safety and Flexibility
+
+To facilitate creating checks for usage dependencies between resources in a workspace and the landing zone. These checks are implemented as DeleteRules. These rules are applied to each resource, and all must pass before starting the delete process.
+
+A rule implements:
+
+```java
+public interface DeleteRule {
+  DeleteRuleResult applyRule(ResourceToDelete genericResource);
+}
+
+```
+
+And to make the implementation of rules simpler, a base class is included with ARM clients and basic checks: `ResourceDependencyDeleteRule`.
+
+A rule then can be implemented focusing on the specific resource that applies, for example a rule that checks if an Azure Relay namespace has hybrid connections.
+
+```java
+public class AzureRelayHasHybridConnections extends ResourceDependencyDeleteRule {
```

Comment:
... same comment re: [understanding the benefits of providing code inline vs directly linking to the code](https://github.com/DataBiosphere/terra-landing-zone-service/blob/5607f8f968492a6691a1aa5cf8d28f50e8d7eaa7/library/src/main/java/bio/terra/landingzone/library/landingzones/management/deleterules/AzureRelayHasHybridConnections.java#L8)